underfs/cephfs/src/main/java/alluxio/underfs/cephfs/CephInputStream.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs.cephfs;
+
+import com.ceph.fs.CephMount;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.InputStream;
+import java.io.IOException;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A stream for reading a file from CephFS.
+ */
+@NotThreadSafe
+public final class CephInputStream extends InputStream {
+  private CephMount mMount;
+  private int mFileHandle;
+  private long mFileLength;
+  private long mPos = 0;
+
+  /** Flag to indicate this stream has been closed, to ensure close is only done once. */
+  private AtomicBoolean mClosed = new AtomicBoolean(false);
+
+  /**
+   * Create a new CephInputStream.
+   * @param mount CephFileSystem mount
+   * @param fh The filehandle provided by Ceph to reference
+   * @param flength The current length of the file. If the length changes
+   *                you will need to close and re-open it to access the new data
+   */
+  public CephInputStream(CephMount mount, int fh, long flength) {
+    mFileLength = flength;
+    mFileHandle = fh;
+    mMount = mount;
+  }
+
+  @Override
+  public int available() throws IOException {
+    return (int) (mFileLength - mPos);
+  }
+
+  /**
+   * Seek.
+   * @param targetPos Position
+   * @throws IOException throws
+   */
+  public void seek(long targetPos) throws IOException {
+    if (targetPos > mFileLength) {
+      throw new IOException(
+          "CephInputStream.seek: failed seek to position " + targetPos
+          + " on fd " + mFileHandle + ": Cannot seek after EOF " + mFileLength);
+    }
+
+    long oldPos = mPos;
+    mPos = mMount.lseek(mFileHandle, targetPos, CephMount.SEEK_SET);
+    if (mPos < 0) {
+      int ret = (int) mPos;
+      mPos = oldPos;
+      throw new IOException("ceph.lseek: ret = " + ret);
+    }
+  }
+
+  @Override
+  public long skip(long n) throws IOException {
+    if (n <= 0) {
+      return 0;
+    }
+    long targetPos = mPos + n;
+    seek(targetPos);
+    return n;
+  }
+
+  @Override
+  public int read() throws IOException {
+    byte[] result = new byte[1];
```

Comment:
this will be created one instance per `read` call.
since this class is marked non-threadsafe, make it a class member?

Reply:
> this will be created one instance per `read` call.
> since this class is marked non-threadsafe, make it a class member?

yes, agree